proxy/src/control/cache.rs

```diff
@@ -93,41 +133,81 @@ impl<T> Cache<T> where T: Clone + Copy + Eq + std::hash::Hash {
         self.reset_on_next_modification = false;
     }
 
-    pub fn clear<F>(&mut self, on_change: &mut F) where F: FnMut(T, CacheChange) {
-        self.retain(&HashSet::new(), on_change)
+    pub fn clear<F>(&mut self, on_change: &mut F)
+    where
+        F: FnMut((K, V), CacheChange),
+    {
+        self.update_intersection(&IndexMap::new(), on_change)
     }
 
-    pub fn retain<F>(&mut self, to_retain: &HashSet<T>, mut on_change: F)
-        where F: FnMut(T, CacheChange)
+    /// Update the cache to contain the intersection of its current contents
+    /// and the key-value pairs in `to_update`. Pairs not present in
+    /// `to_update` will be removed from the cache, and any keys in the cache
+    /// with different values from those in `to_update` will be ovewritten to
+    /// match `to_update`.
+    pub fn update_intersection<F, Q>(
+        &mut self,
+        to_update: &IndexMap<Q, V>,
+        mut on_change: F
+    )
+    where
+        F: FnMut((K, V), CacheChange),
+        K: Borrow<Q>,
+        Q: Hash + Eq,
     {
-        self.values.retain(|value| {
-            let retain = to_retain.contains(&value);
-            if !retain {
-                on_change(*value, CacheChange::Removal)
+        self.values.retain(|key, value| {
+            match to_update.get(key.borrow()) {
+                // New value matches old value. Do nothing.
+                Some(new_value) if new_value == value => true,
+                // If the new value isn't equal to the old value, overwrite
+                // the old value.
+                Some(new_value) =>  {
+                    let old_value = mem::replace(value, new_value.clone());
```

Comment:
Again, we should be able to avoid the clone here, and we should remove the clone in a follow-up PR. (We should be able to make `to_update: &mut IndexMap<...)` and then use `to_update.remove()` instead of `to_update.get()` above.)

Reply:
> to_update: &mut IndexMap<...>

Actually, when we make that change, it should be `mut to_update: IundexMap<...>`; i.e. the function should take ownership of the map so that we don't have to worry about the caller trying to use it for something else after we removed the entries.